### PR TITLE
Refactor backend/qemu 

### PR DESCRIPTION
**Pull request by mudler**
Still WIP, needed the PR to check CI outputs.

As for now:
- [partial refactor of backend::qemu start_qemu ](https://github.com/os-autoinst/os-autoinst/pull/781#issuecomment-300765459) (introduced new functions to improve, hopefully, readability)
- created osutils to export common functions that can be used by other modules as well.
- use in backend::pvm the same functions introduced from the refactor of backend::qemu
- small cleanups in backend::s390x
- prepare arguments before forking qemu in start_qemu ( [Poo#19166](https://progress.opensuse.org/issues/19166) )
- backend: move child process monitor/fork to baseclass (was in backend::qemu)

### CODE REVIEW

**Comment by coolo** [comment]
you don't need a PR - you can have travis on your repo too. But I don't mind

**Comment by mudler** [comment]
`s/needed/also needed/ `😄 - feedback of course is always appreciated

File: backend/qemu.pm [hunk 1]
```diff
@@ -738,26 +701,29 @@ sub start_qemu {
             if ($vars->{VNC} !~ /:/) {
                 $vars->{VNC} = ":$vars->{VNC}";
             }
-            push(@params, "-vnc", "$vars->{VNC},share=force-shared");
-            push(@params, "-k", $vars->{VNCKB}) if ($vars->{VNCKB});
+            gen_params @params, 'vnc', "$vars->{VNC},share=force-shared";
+            gen_params @params, 'k',   $vars->{VNCKB};
```
Comment (mnowaksuse) [inline backend/qemu.pm]: Missing `if ($vars->{VNCKB})`?

Comment (mudler) [inline backend/qemu.pm]: This regards just coverage actually, `gen_params` checks if the variable is empty, and don't push it into the array in that case.

File: backend/s390x.pm [hunk 2]
```diff
@@ -47,7 +45,7 @@ sub do_start_vm {
 
     $self->unlink_crash_file();
     my $console = $testapi::distri->add_console('x3270', 's3270');
-    $console->backend($self);
+    $console->backend($self) if $console;
```
Comment (coolo) [inline backend/s390x.pm]: why would add_console return undef:

Comment (mudler) [inline backend/s390x.pm]: it was meant to make coverage happier :p i'll remove it then

File: backend/pvm.pm [hunk 3]
```diff
@@ -96,30 +97,38 @@ sub pvmctl {
...
     }
     elsif ($type =~ /scsi/) {
         my ($type, $action, $kind, $file, $target) = @_;
         #so far only disks are reatachable to the master LPAR
         #set it back to default if argument is omited
         $lpar = $target if $target;
-        push(@cmd, "--type", $kind, "--stor-id", $file);
-        push(@cmd, "--lpar", "name=" . $lpar);
-        push(@cmd, "--vg", "name=rootvg") if ($type =~ /lv/);
+        dd_gen_params @cmd, "type",    $kind         if $kind;
```
Comment (coolo) [inline backend/pvm.pm]: all these post ifs are very strange. Because the command will not work if one or the other isn't set

Comment (mudler) [inline backend/pvm.pm]: isn't the command gonna die anyway without those parameters? if so i would introduce a guard here

Comment (coolo) [inline backend/pvm.pm]: then die and do not add single parts of it to the command

**Comment by okurz** [comment]
+1

**Comment by mudler** [comment]
Can i proceed to rebase and remove WIP?

**Comment by coolo** [comment]
well, how much testing did this see?

**Comment by okurz** [comment]
you don't need to squash all commits in case they provide something useful on their own

**Comment by coolo** [comment]
yes - the number of commits is reasonable compared to the number of changes.

**Comment by mudler** [comment]
I could test only the qemu backend so far - i do not have s390x or pvm setup right now

**Comment by sysrich** [comment]
@mudler can you give us links to some of your qemu test runs then?

**Comment by mudler** [comment]
Sorry - completely forgot to give links, meanwhile i've run tests again just to be sure. 

Test expected to be **successful**:

- [HDDMODEL=nvme, UEFI=1](http://10.100.51.55/tests/95)  
- [default HDDMODEL and without UEFI](http://10.100.51.55/tests/96)
- [HDDMODEL=virtio-blk, QEMURAM=2048](http://10.100.51.55/tests/99)
- [HDDMODEL=virtio-blk, NUMDISKS=3](http://10.100.51.55/tests/100)
- [HDDMODEL=virtio-blk, QEMU_VIRTIO_RNG=1, UEFI=1](http://10.100.51.55/tests/101)

Test expected to be **unsuccessful** (host machine not configured to run such tests, just to ensure parameters are generated correctly):

- [NICTYPE=tap](http://10.100.51.55/tests/97)
- [NICTYPE=vde](http://10.100.51.55/tests/98)

**Comment by okurz** [comment]
looks quite clean, still WIP?

**Comment by mudler** [comment]
Had to rebase, but tests looks fine

**Comment by coolo** [comment]
+403 −202 isn't really my favorite type of refactoring, but let's see :)

**Pull request merged**